tree/dataframe/inc/ROOT/RDF/RVariation.hxx

```diff
@@ -61,91 +54,103 @@ bool ResultsSizeEq(const RVec<T> &results, std::size_t expected, std::size_t nCo
 }
 
-// This overload is for the case of ret_type == RVec<RVec<...>>
 template <typename T>
-void ResizeResults(ROOT::RVec<ROOT::RVec<T>> &results, std::size_t nCols, std::size_t nVariations)
+std::size_t GetNVariations(const std::vector<RVec<T>> &results)
 {
-   if (nCols == 1) {
-      results.resize(nVariations);
-   } else {
-      results.resize(nCols);
-      for (auto &rvecOverVariations : results) {
-         rvecOverVariations.resize(nVariations);
-      }
-   }
+   assert(!results.empty());
+   return results[0].size();
 }
 
-// Assign into fLastResults[slot] without changing the addresses of its elements (we gave those addresses away in
-// GetValuePtr)
-// This overload is for the case of a single column and ret_type != RVec<RVec<...>>
 template <typename T>
-void AssignResults(ROOT::RVec<T> &resStorage, ROOT::RVec<T> &&tmpResults, std::size_t /*nCols*/)
+void ResizeResults(std::vector<ROOT::RVec<T>> &results, std::size_t nCols, std::size_t nVariations)
 {
-   const auto nVariations = resStorage.size(); // we have already checked that tmpResults has the same size
-
-   for (auto i = 0u; i < nVariations; ++i)
-      resStorage[i] = std::move(tmpResults[i]);
+   results.resize(nCols);
+   for (auto &rvecOverVariations : results)
+      rvecOverVariations.resize(nVariations);
 }
 
-// This overload is for the case of ret_type == RVec<RVec<...>>
 template <typename T>
-void AssignResults(ROOT::RVec<ROOT::RVec<T>> &resStorage, ROOT::RVec<ROOT::RVec<T>> &&tmpResults, std::size_t nCols)
+void AssignResults(std::vector<ROOT::RVec<T>> &resStorage, ROOT::RVec<ROOT::RVec<T>> &&tmpResults)
 {
-   // we have already checked that tmpResults has the same inner size
-   const auto nVariations = nCols == 1 ? resStorage.size() : resStorage[0].size();
-
-   if (nCols == 1) {
+   const auto nCols = resStorage.size();
+   const auto nVariations = resStorage[0].size();
+   for (auto colIdx = 0u; colIdx < nCols; ++colIdx)
       for (auto varIdx = 0u; varIdx < nVariations; ++varIdx)
-         resStorage[varIdx] = std::move(tmpResults[varIdx]);
-   } else {
-      for (auto colIdx = 0u; colIdx < nCols; ++colIdx)
-         for (auto varIdx = 0u; varIdx < nVariations; ++varIdx)
-            resStorage[colIdx][varIdx] = std::move(tmpResults[colIdx][varIdx]);
-   }
+         resStorage[colIdx][varIdx] = std::move(tmpResults[colIdx][varIdx]);
 }
 
+template <typename T>
+void *GetValuePtrHelper(std::vector<ROOT::RVec<T>> &v, std::size_t colIdx, std::size_t varIdx)
+{
+   return static_cast<void *>(&v[colIdx][varIdx]);
+}
+
+/******* End of helper functions *******/
+
+template <typename Ret_t, typename IsSingleColumn = std::true_type>
+struct ColumnType {
+   using type = typename Ret_t::value_type;
+};
+
+template <typename Ret_t>
+struct ColumnType<Ret_t, std::false_type> {
+   using type = typename Ret_t::value_type::value_type;
+};
+
+/// ColumnType_t is Ret_t::value_type if IsSingleColumn is std::true_type, otherwise it is
```

Comment:
Why that's technically correct (according to the code above), it might be more helpful to state this in terms of an example for both cases.

Reply:
done, thanks